pin-depends should reinstall packages which are pinned

If package X depends on package Y and contains a `pin-depends` for Y. If Y is already installed when X is pinned, opam doesn't rebuild Y, even if the user agrees to the pin being added. This can break builds:

```
[dra@ocamldefang logs-syslog]$ opam install syslog-message
The following actions will be performed:
  ∗ install syslog-message 0.0.2

<><> Gathering sources ><><><><><><><><><><><><><><><><><><><><><><><><><><><><>
[syslog-message.0.0.2] found in cache

<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
∗ installed syslog-message.0.0.2
Done.
[dra@ocamldefang logs-syslog]$ opam pin add logs-syslog .
[logs-syslog.0.1.1] synchronised from git+file:///home/dra/logs-syslog#unix-sockets
The following additional pinnings are required by logs-syslog.0.1.1:
  - syslog-message.0.0.3 at git+https://github.com/verbosemode/syslog-message.git#master
Continue? [Y/n] y
[syslog-message.0.0.3] synchronised from git+https://github.com/verbosemode/syslog-message.git#master
syslog-message is now pinned to git+https://github.com/verbosemode/syslog-message.git#master (version 0.0.3)
logs-syslog is now pinned to git+file:///home/dra/logs-syslog#unix-sockets (version 0.1.1)

The following actions will be performed:
  ∗ install logs-syslog 0.1.1*
Do you want to continue? [Y/n] y

<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
[ERROR] The compilation of logs-syslog failed at "/home/dra/.opam/opam-init/hooks/sandbox.sh build ocaml pkg/pkg.ml build --pinned true --with-lwt false --with-lwt-tls false
        --with-mirage false --with-mirage-tls false".

#=== ERROR while compiling logs-syslog.0.1.1 ==================================#
# context     2.0.0 | linux/x86_64 | ocaml-system.4.07.0 | pinned(git+file:///home/dra/logs-syslog#unix-sockets#1827bfa0)
# path        ~/.opam/test/.opam-switch/build/logs-syslog.0.1.1
# command     ~/.opam/opam-init/hooks/sandbox.sh build ocaml pkg/pkg.ml build --pinned true --with-lwt false --with-lwt-tls false --with-mirage false --with-mirage-tls false
# exit-code   1
# env-file    ~/.opam/log/logs-syslog-29228-6aa130.env
# output-file ~/.opam/log/logs-syslog-29228-6aa130.out
### output ###
# [...]
# + ocamlfind ocamlopt -c -g -bin-annot -safe-string -principal -package ptime.clock.os -package 'syslog-message logs ptime' -w +A-4-44 -color always -I src -o src/logs_syslog_unix.cmx src/logs_syslog_unix.ml
# File "src/logs_syslog_unix.ml", line 137, characters 15-42:
# Error: Unbound value Syslog_message.encode_local
# Command exited with code 2.
# pkg.ml: [ERROR] cmd ['ocamlbuild' '-use-ocamlfind' '-classic-display' '-j' '4' '-tag' 'debug'
#      '-build-dir' '_build' 'opam' 'pkg/META' 'CHANGES.md' 'LICENSE.md'
#      'README.md' 'src/logs-syslog.a' 'src/logs-syslog.cmxs'
#      'src/logs-syslog.cmxa' 'src/logs-syslog.cma' 'src/logs_syslog.cmx'y
#      'src/logs_syslog.cmi' 'src/logs_syslog.mli' 'src/logs-syslog-unix.a'
#      'src/logs-syslog-unix.cmxs' 'src/logs-syslog-unix.cmxa'
#      'src/logs-syslog-unix.cma' 'src/logs_syslog_unix.cmx'
#      'src/logs_syslog_unix.cmi' 'src/logs_syslog_unix.mli']: exited with 10



<><> Error report <><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><>
┌─ The following actions failed
│ λ build logs-syslog 0.1.1
└─ 
╶─ No changes have been performed
[NOTE] Pinning command successful, but your installed packages may be out of sync.
```

If Y is not installed when X is pinned, then it works as expected:
```
[dra@ocamldefang logs-syslog]$ opam pin add logs-syslog .
[logs-syslog.0.1.1] synchronised from git+file:///home/dra/logs-syslog#unix-sockets
The following additional pinnings are required by logs-syslog.0.1.1:
  - syslog-message.0.0.3 at git+https://github.com/verbosemode/syslog-message.git#master
Continue? [Y/n] y
[syslog-message.0.0.3] synchronised from git+https://github.com/verbosemode/syslog-message.git#master
syslog-message is now pinned to git+https://github.com/verbosemode/syslog-message.git#master (version 0.0.3)
logs-syslog is now pinned to git+file:///home/dra/logs-syslog#unix-sockets (version 0.1.1)

The following actions will be performed:
  ∗ install syslog-message 0.0.3* [required by logs-syslog]
  ∗ install logs-syslog    0.1.1*
===== ∗ 2 =====
Do you want to continue? [Y/n] y

<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
∗ installed syslog-message.0.0.3
∗ installed logs-syslog.0.1.1
Done.
```

--- a/src/client/opamClient.ml
+++ b/src/client/opamClient.ml
@@ -1262,10 +1262,10 @@ let reinstall t ?(assume_built=false) names =
 module PIN = struct
   open OpamPinCommand
 
-  let post_pin_action st name =
+  let post_pin_action st names =
     try
       upgrade_t ~strict_upgrade:false ~auto_install:true ~ask:true ~all:false
-        [name, None] st
+       (List.map (fun name -> name, None) names) st
     with e ->
       OpamConsole.note
         "Pinning command successful, but your installed packages \
@@ -1293,6 +1293,7 @@ module PIN = struct
 
   let pin st name ?(edit=false) ?version ?(action=true) target =
     try
+      let pinned = st.pinned in
       let st =
         match target with
         | `Source url -> source_pin st name ?version ~edit (Some url)
@@ -1304,13 +1305,19 @@ module PIN = struct
         | `None -> source_pin st name ?version ~edit None
       in
       if action then
-        (OpamConsole.msg "\n"; post_pin_action st name)
+        let names =
+          OpamPackage.Set.Op.(st.pinned -- pinned)
+          |> OpamPackage.Set.elements
+          |> List.map OpamPackage.name
+        in
+        (OpamConsole.msg "\n"; post_pin_action st names)
       else st
     with
     | OpamPinCommand.Aborted -> OpamStd.Sys.exit_because `Aborted
     | OpamPinCommand.Nothing_to_do -> st
 
   let edit st ?(action=true) ?version name =
+    let pinned = st.pinned in
     let st =
       if OpamPackage.has_name st.pinned name then
         edit st ?version name
@@ -1343,7 +1350,14 @@ module PIN = struct
         OpamConsole.error_and_exit `Not_found
           "Package is not pinned, and no existing version was supplied."
     in
-    if action then post_pin_action st name else st
+    if action then
+      let names =
+        OpamPackage.Set.Op.(st.pinned -- pinned)
+        |> OpamPackage.Set.elements
+        |> List.map OpamPackage.name
+      in
+      post_pin_action st names
+    else st
 
   let unpin st ?(action=true) names =
     let pinned_before = st.pinned in

--- a/src/client/opamCommands.ml
+++ b/src/client/opamCommands.ml
@@ -2539,6 +2539,7 @@ let pin ?(unpin_only=false) () =
          in
          OpamGlobalState.with_ `Lock_none @@ fun gt ->
          OpamSwitchState.with_ `Lock_write gt @@ fun st ->
+         let pinned = st.pinned in
          let st =
            List.fold_left (fun st (name, opam_opt) ->
                OpamStd.Option.iter (fun opam ->
@@ -2553,11 +2554,17 @@ let pin ?(unpin_only=false) () =
                   | OpamPinCommand.Nothing_to_do -> st)
              st names
          in
+         (* names are in newly pinned packages *)
+         let atoms =
+           OpamPackage.Set.Op.(st.pinned -- pinned)
+           |> OpamPackage.Set.elements
+           |> List.map (fun p -> (OpamPackage.name p, None))
+         in
          if action then
            let _st =
              OpamClient.upgrade_t
                ~strict_upgrade:false ~auto_install:true ~ask:true ~all:false
-               (List.map (fun (n,_) -> n, None) names) st
+               atoms st
            in
            `Ok ()
          else `Ok ())

